### PR TITLE
Fix #13 - Proposal for processing multiple files and #include directives

### LINKED ISSUE #13: Proposal for processing multiple files and #include directives
#### Multiple input files
```bash
scala-bindgen -C src/generatedSources/scala a.h b.h c.h
```
... generates files a.scala b.scala and c.scala under src/generatedSources/scala

#### Consolidated, single output file

```bash
scala-bindgen -C src/generatedSources/scala -o single.scala a.h b.h c.h
```
... generates a consolidated, single output file src/generatedSources/scala/single.scala

### Processing #include directives

* A header file may include several other header files.
* When generating bindings, only symbols defined on the topmost files (those mentioned in the command line) must have their bindings generated.
* Bindings from the topmost file may reference symbols from included files. In this case, additional output files are generated so that symbols are grouped and organized by their original context. Only those referenced symbols are generated, not the entire included file. For example:

```bash
scala-bindgen topmost.h another.h
```

```scala
object topmost extends topmost
trait topmost {
  @link("topmost")
  def acall: Data = extern
}
```
```scala
object included extends included
trait included {
  case class Data(...)
}
```
```scala
object another extends another
trait another { ... }
```

### PR DESCRIPTION
**Pull request by frgomes**
Attempts to fix #13 

Mimics the behavior or other tools of the Linux ecosystem, like:
* passing -C DIR for saving all output files into DIR;
* Option ``-o`` is now a "unified" output file. This is subject to review and change in future if not convenient.

### CODE REVIEW

**Comment by frgomes** [comment]
@jonas  Please review

File: bindgen/src/main/scala/Main.scala [hunk 1]
```diff
@@ -93,52 +91,98 @@ class Generator(c: Args) {
...
           disposeTranslationUnit(tu)
           result.toInt
         }
       }
     if (index != null) disposeIndex(index)
+    0 //FIXME: obtain from iterator
+  }
+
+  private def makeOutput(tree: Tree, out: String): Unit = {
+    val cout = if("-" == out) new PrintStream(System.out) else new PrintStream(new FileOutputStream(mkdirs(out)))
```
Comment (jonas) [inline bindgen/src/main/scala/Main.scala]: If possible I'd prefer to keep option parsing in the CLI. I understand that opening files up front might not be perfect so it would require some lazy opening.

File: tests/src/test/scala/bindgen/BindgenSpec.scala [hunk 2]
```diff
@@ -1,42 +1,86 @@
...
-        val output = new File(outputDir, expected.getName)
-        val bin = sys.props.get("native.bin").getOrElse {
-          sys.error("native.bin is not set")
-        }
-        val cmd =
-          Array(bin, "-o", output.getAbsolutePath, input.getAbsolutePath)
+import utest._
 
+
+object BindgenSpec extends TestSuite {
```
Comment (jonas) [inline tests/src/test/scala/bindgen/BindgenSpec.scala:6]: Not sure I see the benefit of using utest here. I suggest we leave that to a separate PR.

File: tests/src/test/scala/bindgen/BindgenSpec.scala [hunk 3]
```diff
@@ -1,42 +1,86 @@
...
+  }
+
+  val util = new Object with FileUtils
+
+
+  val tests = this {
+    "ability to generate bindings"-{
+      for {
+        input <- inputDir.listFiles()                                        if input.getName.endsWith(".h")
+        expected = util.resolve(null, null, input.toString, ".h", ".scala")  if (new File(expected)).exists
```
Comment (jonas) [inline tests/src/test/scala/bindgen/BindgenSpec.scala]: Having `null`s here looks look like a code smell.

File: tests/src/test/scala/bindgen/BindgenSpec.scala [hunk 4]
```diff
@@ -1,42 +1,86 @@
...
 
+  def resolve(chdir: String, name: String): String = {
+    val dir  = if(null==chdir || ""==chdir) "." else chdir
+    val path = Paths.get(name)
+    if(path.isAbsolute) path.toString else Paths.get(dir, path.toString).toString
+  }
+
+  def resolve(chdir: String, name: String, default: String): String =
+    (if(name==null) "" else name) match {
+      case "-" => "-"
```
Comment (jonas) [inline tests/src/test/scala/bindgen/BindgenSpec.scala]: Moving parsing of `"-"` to the CLI class would avoid this complexity.

File: tests/src/test/scala/bindgen/BindgenSpec.scala [hunk 5]
```diff
@@ -1,42 +1,86 @@
...
+      case ""  => resolve(chdir, default)
+      case _   => resolve(chdir, name)
+    }
+
+  def resolve(chdir: String, name: String, default: String, from: String, to: String): String =
+    (if(name==null) "" else name) match {
```
Comment (jonas) [inline tests/src/test/scala/bindgen/BindgenSpec.scala]: I'd prefer to use `Option[String]` for `chdir` and `name`.

File: bindgen/src/main/scala/Main.scala [hunk 6]
```diff
@@ -1,19 +1,16 @@
...
-import java.io.{File, FileOutputStream, PrintStream}
 import scalanative.native._
 
-//see: http://bastian.rieck.ru/blog/posts/2015/baby_steps_libclang_ast/
 
 case class Args(files: Seq[String] = Seq(),
                 chdir: String = ".",
                 pkg: String = "scalanative.native.bindings",
-                out: PrintStream = System.out,
+                out: String = "",
```
Comment (jonas) [inline bindgen/src/main/scala/Main.scala]: If we make this `Option[File]` I think we can cover both the case of `"-"` and passing a path.

File: bindgen/src/main/scala/Main.scala [hunk 7]
```diff
@@ -93,52 +91,98 @@ class Generator(c: Args) {
...
-          if (c.debug) println("[about to call visitChildren]")
           val result = visitChildren(root, visitor, tree.cast[Data])
+          //TODO assert(result == CXChildVisit_Continue)
+          makeOutput(tree, resolve(args.chdir, args.out, name, ".h", ".scala"))
           disposeTranslationUnit(tu)
           result.toInt
         }
       }
     if (index != null) disposeIndex(index)
+    0 //FIXME: obtain from iterator
```
Comment (jonas) [inline bindgen/src/main/scala/Main.scala]: BTW, what is the purpose of this return value? Would it be better to bail out if parsing of a translation unit fails?

File: bindgen/src/main/scala/Main.scala (hunk 6 above)
Comment (jonas) [reply bindgen/src/main/scala/Main.scala]: Would also handle the case where the output is a directory and we want to generate multiple files inside it.

Comment (frgomes) [reply bindgen/src/main/scala/Main.scala]: We have 3 cases here (without discussing the types involved):
1. nothing : could be though as "default". The application chooses the output file name according to a number of factors;
2. minus: it means standard output, which Unix takes care for you, whatever it is: a terminal, a file, a pipe, etc.
3. a filename: it can be either a relative file name or an absolute file name.

The type has to be String because the semantic of minus and filename is different. A minus refers to a stream whilst filename refers to a File (or even URL, if you will). The easiest way to cover all cases is simply representing as String.

Yes, it could be Option[String].

File: bindgen/src/main/scala/Main.scala (hunk 7 above)
Comment (frgomes) [reply bindgen/src/main/scala/Main.scala]: The application needs to inform the Unix shell if things went wrong. The point is that I'm not yet plenty sure what should be the criteria to tell that. That's why this thing is still pending at this point.

Conceptually, if we do a ``result = something.map(...).sum`` and it results zero, it means that all steps where zero, which means that everything ran smoothly. But I'm not plenty sure at this point if there are other factors to consider.

File: bindgen/src/main/scala/Main.scala (hunk 6 above)
Comment (jonas) [reply bindgen/src/main/scala/Main.scala]: The only sane default output file I can think of is stdout so we have two cases and ` Option[File]` would handle them

File: bindgen/src/main/scala/Main.scala (hunk 7 above)
Comment (jonas) [reply bindgen/src/main/scala/Main.scala]: Given that this is in the parsing phase and we cannot recover from this failure, it would be OK to write the error to stderr and exit.

File: bindgen/src/main/scala/Main.scala (hunk 6 above)
Comment (frgomes) [reply bindgen/src/main/scala/Main.scala]: Observe that "nothing" is the default, which means that the application logic (which is not responsibility of the CLI layer) is responsible for telling what the actual file is.

Then  you have two cases:
* new File("something.scala"), which is the file the user specified. It's OK.
* new File("-"), which means: a file in the current directory named "-", which is NOT what the user wants. By convention, when the user passes "-" on the command line, it means STDOUT, which is a stream managed by ``/bin/bash``; and not a regular file in the file system named "-".

File: bindgen/src/main/scala/Main.scala (hunk 7 above)
Comment (frgomes) [reply bindgen/src/main/scala/Main.scala]: One factor to consider, for example, is that the order of the input files may be relevant for a successful transformation. I wouldn't like to enforce such behavior. #29

I will change to a fail fast, as suggested.

File: bindgen/src/main/scala/Main.scala (hunk 6 above)
Comment (jonas) [reply bindgen/src/main/scala/Main.scala]: If the CLI parses `-o` using

```scala
if (arg == "-") None else Some(new File(arg))
```

Then the generator can do
```scala
val out = c.out.fold(System.out)(file => new PrintStream(new FileOutputStream(file)))
```

I wouldn't put any logic inside the generator itself other than perhaps handling the case where `c.out` is a directory differently if we want to support outputting to multiple files.

Comment (frgomes) [reply bindgen/src/main/scala/Main.scala]: @jonas :: A fold works, but you are still missing the fact that it's a tri-state thing, and not a dual-state thing. The state "None" must be different from the state when the user enters "-".

This is an example of a tri-state fold:
https://gist.github.com/frgomes/bde451a058efef6448af

Comment (frgomes) [reply bindgen/src/main/scala/Main.scala]: @jonas :: In the example I've provided, there's only one data type, which is Float. In our case here, we have two data types: OutputStream and File. The common denominator is the name we assign to these things. This is why we need Option[String]. Employing Option[File] adds nothing; actually, only complicates things.

Comment (jonas) [reply bindgen/src/main/scala/Main.scala]: Can you explain why it needs to be a tristate?

Comment (jonas) [reply bindgen/src/main/scala/Main.scala]: As I suggested about stdout should be the default state because it won't write anything to the filesystem.

**Comment by frgomes** [comment]
@jonas :: Please have a look again.
* CLI employs ``Option[String]`` now;
* Fixed some smelly calls passing nulls;
* Generator.process now returns sum(errno) from several calls to Clang. It was the most no-brainer thing to be done. Can be changed later, if we find necessary. See: #29

**Comment by frgomes** [comment]
Try this command from the shell:
```bash
$ bindgen/target/scala-2.11/bindgen-out -d -v -C target -o - ~/workspace/scala-bindgen/tests/src/test/resources/samples/test*.h
```

It does not work properly inside SBT because SBT gets confused with a "-" alone in the command line.
See: https://github.com/sbt/sbt/issues/2257

File: bindgen/src/main/scala/Main.scala (hunk 6 above)
Comment (frgomes) [reply bindgen/src/main/scala/Main.scala]: @jonas 

## Tri-state

1. ``None`` (the default). Meaning: assign the output file name(s) according to input file name(s). Notice that one input file may eventually hinge on several output files, according to #23 and #13. It's implemented by something like
```scala
val pw = new PrinterWriter(new FileOutputStream(new File(findOutputFileNameFrom(currentInputFileName))))
```

2. ``Some("/write/all/output/to/this/unified/single/filename.scala")`` : Meaning: the user specifies explicitly that all output must be consolidated into a single output file. It's implemented by something like
```scala
val pw = new PrinterWriter(new FileOutputStream(new File(args.out)))
```

3. ``Some("-")`` : Similar to (2), but writes all output to ``System.out``. It's implemented by something like
```scala
val pw = new PrinterWriter(System.out)
```

## Defaults

Depending on the purpose ot the application, defaults are chosen on different ways.

Tools like compressors and compilers are primarily intended to produce files onto the filesystem.
The same applies to ``scala-bindgen``: it's primarily intended to produce files onto the file system by inferring output file names from input file names, pretty much as compilers do. In regime, we infer output file names; in the exception, the user chooses a specific name or chooses the screen.

Other tools, such as grep or sed, simply cannot assume anything about the output file and so, simply chose to send the output to the screen. In regime, the user prefers the screen; exceptionally, the user redirects the screen to a file.
